auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/external/DeltaSetPasswordController.kt

```diff
@@ -0,0 +1,199 @@
+
+class DeltaSetPasswordController(
+    private val deltaConfig: DeltaConfig,
+    private val ldapConfig: LDAPConfig,
+    private val emailConfig: EmailConfig,
+    private val authServiceConfig: AuthServiceConfig,
+    private val userService: UserService,
+    private val setPasswordTokenService: SetPasswordTokenService,
+    private val userLookupService: UserLookupService,
+    private val emailService: EmailService,
+) {
+    private val logger = LoggerFactory.getLogger(this.javaClass)
+    private val passwordChecker = PasswordChecker()
+
+    fun setPasswordFormRoutes(route: Route) {
+        route.post {
+            setPasswordPost(call)
+        }
+        route.get {
+            setPasswordGet(call)
+        }
+    }
+
+    fun setPasswordSuccessRoute(route: Route) {
+        route.get {
+            call.respondSuccessPage()
+        }
+    }
+
+    private suspend fun setPasswordGet(call: ApplicationCall) {
+        val userCN = call.request.queryParameters["userCN"].orEmpty()
+        val token = call.request.queryParameters["token"].orEmpty()
+        when (setPasswordTokenService.useToken(token, userCN)) {
+            is SetPasswordTokenService.NoSuchToken -> {
+                throw SetPasswordException("set_password_no_token", "Set password token did not exist")
+            }
+
+            is SetPasswordTokenService.ExpiredToken -> {
+                sendNewSetPasswordLink(userCN)
+                call.respondExpiredTokenPage()
+            }
+
+            is SetPasswordTokenService.ValidToken -> {
+                call.respondSetPasswordPage(userCN, token)
+            }
+        }
+    }
+
+    private suspend fun setPasswordPost(call: ApplicationCall) {
+        val formParameters = call.receiveParameters()
+        val userCN = call.request.queryParameters["userCN"].orEmpty()
+        val token = call.request.queryParameters["token"].orEmpty()
+        val newPassword = formParameters["newPassword"].orEmpty()
+        val confirmPassword = formParameters["confirmPassword"].orEmpty()
+
+        if (Strings.isNullOrEmpty(userCN)) throw SetPasswordException(
+            "set_password_no_user_cn",
+            "User CN not present on setting password"
+        )
+
+        if (Strings.isNullOrEmpty(token)) throw SetPasswordException(
+            "set_password_no_token",
+            "Token not present on setting password"
+        )
+
+
+        val message: String? =
+            if (Strings.isNullOrEmpty(newPassword)) "New password is required."
+            else if (Strings.isNullOrEmpty(confirmPassword)) "Confirm password is required."
+            else if (newPassword != confirmPassword) "Passwords did not match."
+            else if (newPassword.length < 12) "Password must be at least 12 characters long."
+            else if (passwordChecker.isCommonPassword(newPassword, userCN))
+                "Passwords must not be a commonly used password format or contain your username"
+            else null
```

Comment:
I think if there are form validation errors let's return an error before checking the token.

I can see why you've done it like this, but I think that would be easier to understand.